src/picterra/base_client.py

```diff
@@ -0,0 +1,264 @@
+        # Create the session with a default timeout (30 sec), that we can then
+        # override on a per-endpoint basis (will be disabled for file uploads and downloads)
+        self.sess = _RequestsSession(timeout=timeout)
+        # Retry: we set the HTTP codes for our throttle (429) plus possible gateway problems (50*),
+        # and for polling methods (GET), as non-idempotent ones should be addressed via idempotency
+        # key mechanism; given the algorithm is {<backoff_factor> * (2 **<retries-1>}, and we
+        # default to 30s for polling and max 30 req/min, the default 5-10-20 sequence should
+        # provide enough room for recovery
+        retry_strategy = Retry(
+            total=max_retries,
+            status_forcelist=[429, 502, 503, 504],
+            backoff_factor=backoff_factor,
+            allowed_methods=["GET"],
+        )
+        adapter = HTTPAdapter(max_retries=retry_strategy)
+        self.sess.mount("https://", adapter)
+        self.sess.mount("http://", adapter)
+        # Authentication
+        self.sess.headers.update({"X-Api-Key": api_key})
+
+    def _full_url(self, path: str, params: dict[str, Any] | None = None):
+        url = urljoin(self.base_url, path)
+        if not params:
+            return url
+        else:
+            qstr = urlencode(params)
+            return "%s?%s" % (url, qstr)
+
+    def _wait_until_operation_completes(
+        self, operation_response: dict[str, Any]
+    ) -> dict[str, Any]:
+        """Polls an operation an returns its data"""
+        operation_id = operation_response["operation_id"]
+        poll_interval = operation_response["poll_interval"]
+        # Just sleep for a short while the first time
+        time.sleep(poll_interval * 0.1)
+        while True:
+            logger.info("Polling operation id %s" % operation_id)
+            resp = self.sess.get(
+                self._full_url("operations/%s/" % operation_id),
+            )
+            if not resp.ok:
+                raise APIError(resp.text)
+            status = resp.json()["status"]
+            logger.info("status=%s" % status)
+            if status == "success":
+                break
+            if status == "failed":
+                errors = resp.json()["errors"]
+                raise APIError(
+                    "Operation %s failed: %s" % (operation_id, json.dumps(errors))
+                )
+            time.sleep(poll_interval)
+        return resp.json()
+
+    def _return_results_page(
+        self, resource_endpoint: str, params: dict[str, Any] | None = None
+    ) -> ResultsPage:
+        if params is None:
+            params = {}
+        if "page_number" not in params:
+            params["page_number"] = 1
+
+        url = self._full_url("%s/" % resource_endpoint, params=params)
+        return ResultsPage(url, self.sess.get)
+
+    def download_operation_results_to_file(self, operation_id: str, filename: str):
```

Comment:
(I know you inherited this from my branch). I wasn't sure about leaving the operations stuff in the base class, even though both APIs work with this. It doesn't seem quite right for a base class to be concerned with these operations, but on the other hand if we add another product that doesn't have an operations endpoint or we have reason to diverge the two endpoints then I think we could easily move these methods later. WDYT?

Reply:
We can always split this later indeed. For now I think it doesn't hurt to keep these Operations concerns in the base class.

I'll just double-check usages though, because the implem of `PlotsAnalysisClient` here doesn't use `download_operation_results_to_file` or any of the operations helpers (which it probably should).

Reply:
Removed a couple of unused operations-related methods that were indeed unused.

Reply:
Thanks :)